calour/io.py

```diff
@@ -103,23 +144,58 @@ def _read_qiime2(fp, transpose=True):
+        oname = _file_from_zip(tempdir, fp, internal_data='data/feature-table.biom')
+        sid, fid, data, fmd = _read_biom(oname, transpose=transpose)
 
-    sid = table.ids(axis='sample')
-    fid = table.ids(axis='observation')
-    logger.info('loaded %d samples, %d observations' % (len(sid), len(fid)))
-    data = table.matrix_data
-    feature_md = _get_md_from_biom(table)
+    return sid, fid, data, fmd
 
-    if transpose:
-        logger.debug('transposing table')
-        data = data.transpose()
 
-    return sid, fid, data, feature_md
+def read_qiime2(fp, sample_metadata_file=None, rep_seq_file=None, taxonomy_file=None, use_qiime2=False, **kwargs):
+    '''Read a qiime2 feature table and additional optional artifact files (representative sequences and taxonomy) into a Calour.AmpliconExperiment
+
+    Parameters
+    ----------
+    fp: str
+        name of the qiime2 feature table .qza artifact file
+    sample_metadata_file : None or str, optional
+        None (default) to just use sample names (no additional metadata).
+        if not None, file path to the sample metadata (aka mapping file in QIIME).
+    rep_seq_file: None or str, optional
+        None (default) to use the feature ids in the feature table
+        if not None, file path to the qiime2 representative sequences artifact file (defined by the qiime2 --o-representative-sequences parameter)
+    taxonomy_file: None or str, optional
+        if not None, add taxonomy for each feature using the qiime2 taxonomy artifact file (output of the qiime2 feature-classifier command)
+
+    Keyword Arguments
+    -----------------
+    %(io.read.parameters)s
+    '''
+    newexp = read_amplicon(fp, sample_metadata_file=sample_metadata_file, data_file_type='qiime2', **kwargs)
+    with tempfile.TemporaryDirectory() as tempdir:
+        # if rep-seqs file is supplied, translate hashes to sequences
+        if rep_seq_file is not None:
+            logger.debug('loading rep_seqs file %s' % rep_seq_file)
+            rep_seqs = []
+            rs_name = _file_from_zip(tempdir, rep_seq_file, internal_data='data/dna-sequences.fasta')
+            for cseq in skbio.read(rs_name, format='fasta'):
+                cseq = str(cseq).upper()
+                rep_seqs.append(cseq)
+            newexp.feature_metadata['_hash'] = newexp.feature_metadata['_feature_id']
+            newexp.feature_metadata['_feature_id'] = rep_seqs
+            newexp.feature_metadata.set_index('_feature_id', inplace=True)
+
+        # if taxonomy file is supplied, load it into the feature metadata
+        if taxonomy_file is not None:
+            logger.debug('loading taxonomy file %s' % taxonomy_file)
+            tax_name = _file_from_zip(tempdir, taxonomy_file, internal_data='data/taxonomy.tsv')
+            taxonomy_df = pd.read_table(tax_name)
+            taxonomy_df.set_index('Feature ID', inplace=True)
+            newexp.feature_metadata = pd.concat([newexp.feature_metadata, taxonomy_df], axis=1)
```

Comment:
does concat align row index? in case it is not matched.

Reply:
yup. see for example:
a=pd.DataFrame({'a':[1,2,3]},index=[4,5,6])
b=pd.DataFrame({'aa':[5,6,4]},index=[6,5,4])
c=pd.concat([a,b],axis=1)
print(c)


   a  aa
4  1   4
5  2   6
6  3   5